Fix flaky test_barrier_timeout_global.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57523 Fix flaky test_barrier_timeout_global.**

`_test_barrier_timeout` would run a barrier on rank 1 and sleep for
`timeout` on other ranks. In some cases if the other ranks would be faster,
they would enter the sleep call much earlier than rank 0 would enter barrier.
As a result, they would exit before the timeout is up and rank 0 would receive
a connection closed error instead of a timeout error. This would result in the
barrier call exiting before the timeout and the subsequent assertion failing.

#Closes: https://github.com/pytorch/pytorch/issues/57176

Differential Revision: [D28170821](https://our.internmc.facebook.com/intern/diff/D28170821/)

pritamdamania committed:
Fix flaky test_barrier_timeout_global.

`_test_barrier_timeout` would run a barrier on rank 1 and sleep for
`timeout` on other ranks. In some cases if the other ranks would be faster,
they would enter the sleep call much earlier than rank 0 would enter barrier.
As a result, they would exit before the timeout is up and rank 0 would receive
a connection closed error instead of a timeout error. This would result in the
barrier call exiting before the timeout and the subsequent assertion failing.

#Closes: https://github.com/pytorch/pytorch/issues/57176

Differential Revision: [D28170821](https://our.internmc.facebook.com/intern/diff/D28170821/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM. We could also consider doing a `dist.barrier()` right before the logic for the ranks diverges so that there's less of a chance skewing significantly and causing flakiness, though this should be good enough as well.

pritamdamania committed:
Update on "Fix flaky test_barrier_timeout_global."


`_test_barrier_timeout` would run a barrier on rank 1 and sleep for
`timeout` on other ranks. In some cases if the other ranks would be faster,
they would enter the sleep call much earlier than rank 0 would enter barrier.
As a result, they would exit before the timeout is up and rank 0 would receive
a connection closed error instead of a timeout error. This would result in the
barrier call exiting before the timeout and the subsequent assertion failing.

#Closes: https://github.com/pytorch/pytorch/issues/57176

Differential Revision: [D28170821](https://our.internmc.facebook.com/intern/diff/D28170821/)

[ghstack-poisoned]
